src/CabanaPD_HeatTransfer.hpp

```diff
@@ -0,0 +1,127 @@
+        std::is_same_v<typename model_type::fracture_type, Elastic> );
+
+    Timer _euler_timer = base_type::_energy_timer;
+    ModelType _model;
+
+  public:
+    HeatTransfer( const bool half_neigh, const ModelType model )
+        : base_type( half_neigh )
+        , _model( model )
+    {
+    }
+
+    template <class TemperatureType, class PosType, class ParticleType,
+              class NeighListType, class ParallelType>
+    void
+    computeHeatTransferFull( TemperatureType& conduction, const PosType& x,
+                             const PosType& u, const ParticleType& particles,
+                             const NeighListType& neigh_list, const int n_local,
+                             ParallelType& neigh_op_tag )
+    {
+        _timer.start();
+
+        auto model = _model;
+        const auto vol = particles.sliceVolume();
+        const auto temp = particles.sliceTemperature();
+
+        auto temp_func = KOKKOS_LAMBDA( const int i, const int j )
+        {
+            double xi, r, s;
+            getDistance( x, u, i, j, xi, r, s );
+
+            const double coeff = model.microconductivity_function( xi );
+            conduction( i ) +=
+                coeff * ( temp( j ) - temp( i ) ) / xi / xi * vol( j );
+        };
+
+        Kokkos::RangePolicy<ExecutionSpace> policy( 0, n_local );
+        Cabana::neighbor_parallel_for(
+            policy, temp_func, neigh_list, Cabana::FirstNeighborsTag(),
+            neigh_op_tag, "CabanaPD::HeatTransfer::computeFull" );
+
+        _timer.stop();
+    }
+
+    template <class ParticleType>
+    void forwardEuler( const ParticleType& particles, const double dt )
+    {
+        _euler_timer.start();
+        auto model = _model;
+        const auto rho = particles.sliceDensity();
+        const auto conduction = particles.sliceTemperatureConduction();
+        auto temp = particles.sliceTemperature();
+        auto n_local = particles.n_local;
+        auto euler_func = KOKKOS_LAMBDA( const int i )
+        {
+            temp( i ) += dt / rho( i ) / model.cp * conduction( i );
+        };
+        Kokkos::RangePolicy<ExecutionSpace> policy( 0, n_local );
+        Kokkos::parallel_for( "CabanaPD::HeatTransfer::forwardEuler", policy,
+                              euler_func );
+        _euler_timer.stop();
+    }
+};
+
+// Heat transfer free function.
+template <class HeatTransferType, class ParticleType, class NeighListType,
+          class ParallelType>
+void computeHeatTransfer( HeatTransferType& heat_transfer,
+                          ParticleType& particles,
+                          const NeighListType& neigh_list,
+                          const ParallelType& neigh_op_tag, const double dt )
+{
+    auto n_local = particles.n_local;
+    auto x = particles.sliceReferencePosition();
+    auto u = particles.sliceDisplacement();
+    auto conduction = particles.sliceTemperatureConduction();
+    auto conduction_a = particles.sliceTemperatureConductionAtomic();
```

Comment:
Could you please remind me the difference between "sliceTemperatureConduction" and "sliceTemperatureConductionAtomic"?

Reply:
atomic is for guarding against race conditions